src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
 
         None
     }
-}
 
-impl From<Options> for Folder {
-    fn from(opt: Options) -> Self {
-        Self {
-            stack: VecDeque::default(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            cache_inlines: Vec::new(),
-            opt,
-            stack_str_size: 0,
-        }
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        assert!(self.occurrences.is_concurrent());
+        assert!(self.opt.nstacks_per_job > 0);
+        assert!(self.opt.nthreads > 1);
+
+        enum Message {
```

Comment:
I think I'd just use `Option<Vec<u8>>` here. And then use `None` to indicate EOF, like with `Iterator`.